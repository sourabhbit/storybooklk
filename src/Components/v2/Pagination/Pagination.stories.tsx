import { ComponentMeta, ComponentStory } from '@storybook/react'
import { ComponentSizeENUM, ThemeENUM, TypographyENUM } from 'Types/general'
import { generateArgTypes, generateGeneralArgs } from 'Helpers/Storybook'
import Pagination from './index'
import { PaginationTypesENUM } from './Pagination.types'

export default {
	title: 'Components/v2/Pagination',
	component: Pagination,
	argTypes: {
		...generateGeneralArgs(),
		type: generateArgTypes({
			type: 'string',
			required: true,
			description: 'Type of Pagination',
			category: 'Component',
			defaultValue: PaginationTypesENUM.storyMode,
			options: PaginationTypesENUM,
		}),
		activeIndex: generateArgTypes({
			type: 'number',
			required: true,
			description: 'active page',
			category: 'Component',
		}),
		count: generateArgTypes({
			type: 'number',
			required: true,
			description: 'Show/Hide Count',
			category: 'Component',
		}),
		onClick: generateArgTypes({
			type: 'function',
			required: false,
			description: 'onClick handler for filter',
			category: 'Events',
		}),
	},
} as ComponentMeta<typeof Pagination>

const Template: ComponentStory<typeof Pagination> = (args) => (
	<Pagination {...args} />
)

export const Image = Template.bind({})
Image.args = {
	id: 'pagination-image-id',
	type: PaginationTypesENUM.image,
	count: 10,
	activeIndex: 1,
	componentSize: ComponentSizeENUM.small,
	theme: ThemeENUM.primary,
	font: TypographyENUM.serif,
}

export const Video = Template.bind({})
Video.args = {
	id: 'pagination-video-id',
	type: PaginationTypesENUM.video,
	count: 12,
	activeIndex: 1,
	componentSize: ComponentSizeENUM.small,
	theme: ThemeENUM.primary,
	font: TypographyENUM.serif,
}

export const StoryMode = Template.bind({})
StoryMode.args = {
	id: 'pagination-pdp-id',
	type: PaginationTypesENUM.storyMode,
	count: 12,
	activeIndex: 1,
	componentSize: ComponentSizeENUM.small,
	theme: ThemeENUM.primary,
	font: TypographyENUM.serif,
}
