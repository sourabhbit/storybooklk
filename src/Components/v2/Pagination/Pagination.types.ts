import { ComponentSizeENUM, ThemeENUM, TypographyENUM } from 'Types/general'
export interface PaginationTypes {
	id: string
	type: PaginationTypesENUM
	activeIndex: number
	theme: ThemeENUM
	font: TypographyENUM
	componentSize?: ComponentSizeENUM
	count: number
	onClick?: (...args: number[]) => number
}

export enum PaginationTypesENUM {
	image = 'image',
	video = 'video',
	storyMode = 'storyMode',
}

export enum PositionENUM {
	left = 'left',
	right = 'right',
}

export enum DimensionENUM {
	height = 'height',
	width = 'width',
}
