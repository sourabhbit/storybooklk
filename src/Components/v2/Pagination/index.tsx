import {
	PaginationWrapper,
	PaginationCircle,
	SemiCircleContainer,
	PaginationLine,
} from './styles'
import {
	PaginationTypes,
	PaginationTypesENUM,
	PositionENUM,
} from './Pagination.types'
import { ComponentSizeENUM, ThemeENUM, TypographyENUM } from 'Types/general'

/**
 * Default Pagination Component
 */
const Pagination = ({
	id,
	type = PaginationTypesENUM.storyMode,
	componentSize = ComponentSizeENUM.medium,
	activeIndex = 1,
	count = 10,
	theme = ThemeENUM.primary,
	font = TypographyENUM.serif,
	onClick,
}: PaginationTypes) => {
	const modIndex = count !== 0 ? activeIndex % count : 0
	return (
		<PaginationWrapper
			data-testid="pagination-wrapper"
			styleFont={font}
			styleSize={componentSize}
			id={id}
		>
			{type === PaginationTypesENUM.image
				? Array.from({ length: count + 2 }, (elem, idx) => (
						<SemiCircleContainer
							theme={theme}
							size={componentSize}
							data-testid="semi-circle-container"
							key={idx}
							active={
								idx === modIndex + 1 && idx !== 0 && idx !== count + 1
							}
							onClick={() => {
								if (
									idx !== 0 &&
									idx !== count + 1 &&
									onClick &&
									typeof onClick === 'function'
								)
									onClick(idx - 1)
							}}
						>
							{idx !== 0 && (
								<PaginationCircle
									position={PositionENUM.left}
									size={componentSize}
									data-testid="pagination-circle-left"
									active={idx === modIndex + 1 || idx === modIndex + 2}
									theme={theme}
								/>
							)}
							{idx !== count + 1 && (
								<PaginationCircle
									position={PositionENUM.right}
									size={componentSize}
									data-testid="pagination-circle-right"
									active={idx === modIndex + 1 || idx === modIndex}
									theme={theme}
								/>
							)}
						</SemiCircleContainer>
				  ))
				: Array.from({ length: count }, (elem, idx) => (
						<PaginationLine
							active={idx === modIndex}
							key={idx}
							data-testid="pagination-line"
							size={componentSize}
							theme={theme}
							lineType={type}
							onClick={() => {
								if (onClick && typeof onClick === 'function')
									onClick(idx)
							}}
						/>
				  ))}
		</PaginationWrapper>
	)
}

export default Pagination
