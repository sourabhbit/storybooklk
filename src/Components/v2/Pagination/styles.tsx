import styled from 'styled-components'
import { ComponentSizeENUM, ThemeENUM, TypographyENUM } from 'Types/general'
import {
	PositionENUM,
	PaginationTypesENUM,
	DimensionENUM,
} from './Pagination.types'

const colorPalette = {
	darkBlue100: 'var(--dark-blue-100)',
	inherit: 'var(--inherit)',
	white: 'var(--white)',
}
const fontPalette = {
	small: '--fs-12',
	medium: '--fs-16',
	large: '--fs-20',
}
const sizePalette = (size: ComponentSizeENUM) => {
	switch (size) {
		case ComponentSizeENUM.small:
			return {
				circleSize: '6px',
				lineMargin: {
					[PaginationTypesENUM.video]: '4px',
					[PaginationTypesENUM.storyMode]: '13px',
				},
				lineHeight: '2px',
				lineWidth: {
					normal: '50px',
					active: '130px',
				},
			}
			break
		case ComponentSizeENUM.medium:
			return {
				circleSize: '9px',
				lineMargin: {
					[PaginationTypesENUM.video]: '6px',
					[PaginationTypesENUM.storyMode]: '19.5px',
				},
				lineHeight: '3px',
				lineWidth: {
					normal: '75px',
					active: '195px',
				},
			}
			break
		case ComponentSizeENUM.large:
			return {
				circleSize: '12px',
				lineMargin: {
					[PaginationTypesENUM.video]: '8px',
					[PaginationTypesENUM.storyMode]: '26px',
				},
				lineHeight: '4px',
				lineWidth: {
					normal: '100px',
					active: '260px',
				},
			}
			break
		default:
			return {
				circleSize: '9px',
				lineMargin: {
					[PaginationTypesENUM.video]: '6px',
					[PaginationTypesENUM.storyMode]: '19.5px',
				},
				lineHeight: '3px',
				lineWidth: {
					normal: '75px',
					active: '195px',
				},
			}
			break
	}
}
const getBorderColor = (colorType: ThemeENUM) => {
	if (colorType === ThemeENUM.primary) {
		return colorPalette.darkBlue100
	} else {
		return colorPalette.white
	}
}

const getBGColor = (
	colorType: ThemeENUM,
	active: boolean,
	type: PaginationTypesENUM
) => {
	if (type === PaginationTypesENUM.image) {
		if (colorType === ThemeENUM.primary) {
			if (active) {
				return colorPalette.darkBlue100
			} else {
				return colorPalette.inherit
			}
		} else {
			if (active) {
				return colorPalette.white
			} else {
				return colorPalette.inherit
			}
		}
	} else {
		if (colorType === ThemeENUM.primary) {
			return colorPalette.darkBlue100
		} else {
			return colorPalette.white
		}
	}
}

const getPosition = (
	position: PositionENUM,
	expectedPosition: PositionENUM,
	size: ComponentSizeENUM
) => {
	if (
		position === PositionENUM.left &&
		expectedPosition === PositionENUM.left
	) {
		return `calc(-${sizePalette(size).circleSize} / 2)`
	} else if (
		position === PositionENUM.right &&
		expectedPosition === PositionENUM.right
	) {
		return `calc(-${sizePalette(size).circleSize} / 2)`
	} else {
		return 'auto'
	}
}

const getRightMargin = (
	lineType: PaginationTypesENUM,
	size: ComponentSizeENUM
) => {
	if (lineType === PaginationTypesENUM.video) {
		return sizePalette(size).lineMargin[lineType]
	} else {
		return sizePalette(size).lineMargin[PaginationTypesENUM.storyMode]
	}
}

const getLineDimensions = (
	lineType: PaginationTypesENUM,
	active: boolean,
	dimensionType: DimensionENUM,
	size: ComponentSizeENUM
) => {
	if (lineType === PaginationTypesENUM.video) {
		if (dimensionType === DimensionENUM.height) {
			if (active) return sizePalette(size).lineHeight
			else return `calc(${sizePalette(size).lineHeight} / 2)`
		} else {
			if (active) return sizePalette(size).lineWidth.active
			else return sizePalette(size).lineWidth.normal
		}
	} else {
		if (dimensionType === DimensionENUM.height)
			return sizePalette(size).lineHeight
		else return sizePalette(size).lineWidth.normal
	}
}

export const PaginationWrapper = styled.div<{
	styleSize: ComponentSizeENUM
	styleFont: TypographyENUM
}>`
	display: flex;
	justify-content: space-between;
	max-width: fit-content;
	background-color: inherit;
	align-items: flex-end;
	font-size: ${(props) => fontPalette[props.styleSize]};
	font-family: ${(props) => props.styleFont};
`
export const SemiCircleContainer = styled.div<{
	theme: ThemeENUM
	active: boolean
	size: ComponentSizeENUM
}>`
	display: flex;
	position: relative;
	width: calc(3 * ${(props) => sizePalette(props.size).circleSize});
	height: ${(props) => sizePalette(props.size).circleSize};
	background-color: ${(props) =>
		getBGColor(props.theme, props.active, PaginationTypesENUM.image)};
	overflow: hidden;
	transition: background-color 300ms ease-in-out;
	cursor: pointer;
`
export const PaginationCircle = styled.div<{
	position: PositionENUM
	theme: ThemeENUM
	active: boolean
	size: ComponentSizeENUM
}>`
	border-radius: 50%;
	border: calc(${(props) => sizePalette(props.size).circleSize} / 6) solid
		${(props) => getBorderColor(props.theme)};
	background-color: ${(props) =>
		getBGColor(props.theme, props.active, PaginationTypesENUM.image)};
	width: ${(props) => sizePalette(props.size).circleSize};
	height: ${(props) => sizePalette(props.size).circleSize};
	box-sizing: border-box;
	position: absolute;
	right: ${(props) =>
		getPosition(props.position, PositionENUM.right, props.size)};
	left: ${(props) =>
		getPosition(props.position, PositionENUM.left, props.size)};
	top: 0;
	z-index: 1;
	transition: background-color 300ms ease-in-out, border 100ms ease-in-out;
`
export const PaginationLine = styled.div<{
	theme: ThemeENUM
	lineType: PaginationTypesENUM
	active: boolean
	size: ComponentSizeENUM
}>`
	display: block;
	margin-right: ${(props) => getRightMargin(props.lineType, props.size)};
	height: ${(props) =>
		getLineDimensions(
			props.lineType,
			props.active,
			DimensionENUM.height,
			props.size
		)};
	width: ${(props) =>
		getLineDimensions(
			props.lineType,
			props.active,
			DimensionENUM.width,
			props.size
		)};
	background-color: ${(props) =>
		getBGColor(props.theme, props.active, props.lineType)};
	opacity: ${(props) => (props.active ? '1' : '0.2')};
	transition: opacity 300ms ease-in-out, width 300ms ease-in-out,
		height 300ms ease-in-out;
	cursor: pointer;
`
