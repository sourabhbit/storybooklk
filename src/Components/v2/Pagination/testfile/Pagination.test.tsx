import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import {
	ComponentSizeENUM,
	ThemeENUM,
	TypographyENUM,
} from '../../../../Types/general'
import Pagination from '../index'
import { PaginationTypesENUM } from '../Pagination.types'

const props = {
	id: 'pagination-id',
	type: PaginationTypesENUM.storyMode,
	componentSize: ComponentSizeENUM.medium,
	activeIndex: 1,
	count: 10,
	theme: ThemeENUM.primary,
	font: TypographyENUM.serif,
	onClick: jest.fn(),
}

describe('<Pagination />', () => {
	let _asFragment: any = null
	let _rerender: any = null

	beforeEach(() => {
		const { asFragment, rerender } = render(<Pagination {...props} />)
		_asFragment = asFragment
		_rerender = rerender
	})

	afterAll(() => {
		_asFragment = null
	})

	it('Should render component', () => {
		const container = screen.getByTestId('pagination-wrapper')
		expect(container).toBeInTheDocument()
	})

	it('Should render pdp/video Component', () => {
		const count = 1
		_rerender(
			<Pagination
				{...props}
				count={count}
				type={PaginationTypesENUM.storyMode}
			/>
		)
		const paginationLine = screen.queryAllByTestId('pagination-line')
		expect(paginationLine[0]).toBeInTheDocument()
		expect(paginationLine.length === count)
	})

	it('Should render circle component', () => {
		const count = 1
		_rerender(
			<Pagination
				{...props}
				count={count}
				type={PaginationTypesENUM.image}
			/>
		)
		const stepCounts = screen.queryAllByTestId('semi-circle-container')
		const leftCircles = screen.queryAllByTestId('pagination-circle-left')
		const rightCircles = screen.queryAllByTestId('pagination-circle-right')
		expect(stepCounts[0]).toBeInTheDocument()
		expect(stepCounts.length === count + 2)
		expect(leftCircles[0]).toBeInTheDocument()
		expect(rightCircles[0]).toBeInTheDocument()
		expect(
			leftCircles.length === count + 1 && rightCircles.length === count + 1
		)
	})

	it('Should take a snapshot', () => {
		expect(_asFragment()).toMatchSnapshot()
	})
})
