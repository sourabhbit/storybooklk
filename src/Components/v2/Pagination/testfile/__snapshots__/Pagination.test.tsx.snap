// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL bNncAS"
    data-testid="pagination-wrapper"
    id="pagination-id"
  >
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye rOYsa"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
    <div
      class="sc-hKwDye jrXdCv"
      data-testid="pagination-line"
    />
  </div>
</DocumentFragment>
`;
